mypy/build.py

```diff
@@ -912,16 +950,11 @@ def write_cache(id: str, path: str, tree: MypyFile,
         manager.trace("Interface for {} is unchanged".format(id))
     else:
         manager.trace("Interface for {} has changed".format(id))
-        try:
-            with open(data_json_tmp, 'w') as f:
-                f.write(data_str)
-                f.write('\n')
-            os.replace(data_json_tmp, data_json)
-            data_mtime = os.path.getmtime(data_json)
-        except os.error as err:
+        data_str += '\n'  # Fast in CPython (it does this in-place if len(data_str) under 10^6)
```

Comment:
I refactored that, because I found out that it's [very unpredictable](https://stackoverflow.com/questions/44487537/why-does-naive-string-concatenation-become-quadratic-above-a-certain-length) when this optimization kicks in (what I thought was a ~10**6 cutoff turned out to be OS-dependent zone where `realloc` starts to fail more and more often).